### PR TITLE
[Feature] Split cross-entropy computation in SP

### PR DESCRIPTION
**Pull request by Edenzzzz**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
1. Avoid gathering hidden states before cross entropy computation to save comm & compute. All-reduce the final loss & grad instead.
2.  Support SP for GPT
3. Merge the logic of PP model fwd and model fwd for GPT2 and Llama for easy maintainence in the future.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/attn.py [hunk 1]
```diff
@@ -890,6 +889,7 @@ def backward(ctx, dout, _):
 
         local_sp_rank = dist.get_rank(sp_group)
         sp_size = dist.get_world_size(sp_group)
+        dist.get_rank(sp_group)
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:892]: meaningless?

File: colossalai/shardformer/modeling/qwen2.py [hunk 2]
```diff
@@ -537,8 +534,10 @@ def forward(
         # Because the input can be padded, the absolute sequence length depends on the max position id.
         rotary_seq_len = max(kv_seq_len, position_ids[:, -1].max().item()) + 1
         cos, sin = self.rotary_emb(value_states, seq_len=rotary_seq_len)
-
-        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+        try:
+            query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+        except Exception as e:
+            raise e
```
Comment (ver217) [inline colossalai/shardformer/modeling/qwen2.py:540]: Useless try-catch?

File: colossalai/shardformer/policies/llama.py [hunk 3]
```diff
@@ -77,6 +77,11 @@ def module_policy(self) -> Dict[Union[str, nn.Module], ModulePolicyDescription]:
         num_q_heads = self.model.config.num_attention_heads
         num_kv_heads = getattr(self.model.config, "num_key_value_heads", None)
 
+        tp_size = self.shard_config.tensor_parallel_size
+        # Modified by SP and TP
+        num_q_heads = self.model.config.num_attention_heads
+        num_kv_heads = getattr(self.model.config, "num_key_value_heads", None)
```
Comment (ver217) [inline colossalai/shardformer/policies/llama.py:83]: Duplicated

Comment (Edenzzzz) [reply colossalai/shardformer/policies/llama.py]: removed

File: colossalai/shardformer/layer/_operation.py [hunk 4]
```diff
@@ -1097,13 +1097,16 @@ def all_to_all_comm(input_, process_group=None, scatter_dim=2, gather_dim=1, fp8
     return _AllToAll.apply(input_, process_group, scatter_dim, gather_dim, fp8_communication)
 
 
-def gather_sp_output(hidden_states, sp_group, sp_mode, fp8_communication=False):
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:1100]: keep fp8 option

File: colossalai/shardformer/layer/loss.py [hunk 5]
```diff
@@ -226,13 +225,13 @@ def dist_cross_entropy(
             logits,
             labels,
             process_group=shard_config.tensor_parallel_process_group,
-            vocab_size=out_features,
+            vocab_size=vocab_size,
```
Comment (ver217) [inline colossalai/shardformer/layer/loss.py:228]: As we may pad lm head, vocab size is not always equal to out features. I doubt whether we can use only one argument here.

Comment (Edenzzzz) [reply colossalai/shardformer/layer/loss.py:228]: In my experience, this works for all models currently in shardformer. And we're actually passing in `lm_head.out_features`?

File: colossalai/shardformer/layer/_operation.py (hunk 4 above)
Comment (Edenzzzz) [reply colossalai/shardformer/layer/_operation.py:1100]: fixed

File: colossalai/shardformer/layer/loss.py (hunk 5 above)
Comment (ver217) [reply colossalai/shardformer/layer/loss.py:228]: @flybird11111 Can you take a look?

Comment (flybird11111) [reply colossalai/shardformer/layer/loss.py:228]: > In my experience, this works for all models currently in shardformer, since we're actually passing in `lm_head.out_features`?

Hi, lm.out_features is the filled vocabulary size. When the weight of lm_head is filled, it is not equal to config.vocabe_size at all. The vocabulary_size of the test_kit model can be divisible by 64 and does not require padding, so there is no problem.

Comment (Edenzzzz) [reply colossalai/shardformer/layer/loss.py:228]: In any case, this won't change the loss since no tokens will be mapped to padded ids?

Comment (flybird11111) [reply colossalai/shardformer/layer/loss.py:228]: > In any case, this won't change the loss since no tokens will be mapped to padded ids?

The padded vocabulary size is used to calculate the vocabulary size per partition. If the last rank uses logit.size[-1] for the vocabulary size per partition, it will result in an error.

Comment (Edenzzzz) [reply colossalai/shardformer/layer/loss.py:228]: I think the padded tokens won't contribute to the loss. I've tried chaging llama's vocab size to 32001, and the tests worked fine.

Comment (flybird11111) [reply colossalai/shardformer/layer/loss.py:228]: > In any case, this won't change the loss since no tokens will be mapped to padded ids?

Because the output of the last rank is the unpadded size.
image

**Review by flybird11111 — APPROVED** [review]

File: colossalai/shardformer/layer/loss.py (hunk 5 above)
Comment (flybird11111) [reply colossalai/shardformer/layer/loss.py:228]: Sorry, you provided the correct padded vocabulary size. It's okay.

**Comment by Edenzzzz** [comment]
@ver217 Ready to merge. Thx

**Review by ver217 — APPROVED** [review]
